File: iOS/Objc/Objc/SignMessage.m
//
//  SignMessage.m
//  Objc
//
//  Created by Shilei Mao on 04/08/2021.
//

#import "SignMessage.h"
#import <HorizonFintexSDK/HorizonFintexSDK.h>

@interface SignMessage ()
@property (strong, nonatomic) id<EthereumOperator> ethOperator;
@end

@implementation SignMessage

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.ethOperator = [EthereumCreator createOperatorWithAttachedView:self.view];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)btSignMessageTapped:(id)sender {
    NSString *keystoreStr = [[self.tfKeystore text] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if ([self isEmptyStr:keystoreStr]) {
        [self showMessageWithTitle:@"Error" message:@"Please enter keystore"];
        return;
    }
    
    NSString *pwdStr = [self.tfPwd text];
    if ([self isEmptyStr:pwdStr]) {
        
        [self showMessageWithTitle:@"Error" message:@"Please enter password"];
        return;
    }
    
    NSString *messageToSign = [self.tfMessage text];
    if ([self isEmptyStr:messageToSign]) {
        [self showMessageWithTitle:@"Error" message:@"Please enter message"];
        return;
    }
    
    NSString *contractAddress = [[self.tfContractAddr text] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if ([self isEmptyStr:contractAddress]) {
        [self showMessageWithTitle:@"Error" message:@"Please enter contract addr"];
        return;
    }
    
    NSString  * _Nullable keystore = [[EthereumKeystore alloc] initWithKeystoreJSON:keystoreStr];
    if (keystore == nil) {
        [self showMessageWithTitle:@"Error" message:@"You've entered invalid keystore, please try again"];
        return;
    }
    
    [self showProgress];
    [[self ethOperator] signMessageWithContractAddr:contractAddress keystore:keystore password:pwdStr signOriginal:messageToSign completion:^(NSString * _Nullable result, NSError * _Nullable error) {
        [self dismissProgress];
        
        if (error != nil) {
            [self showMessageWithTitle:@"Error" message:[error description]];
            return;
        }
        
        [self showMessageWithTitle:@"Signed message" message:result];
    }];
    
}

- (void)showMessageWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alertView = [UIAlertController
                                    alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil];
    
    [alertView addAction:cancelAction];
    
    [self presentViewController:alertView animated:true completion:nil];
}

- (void)showProgress {
    UIViewController *progressView = [self.storyboard instantiateViewControllerWithIdentifier:@"ProgressView"];
    [progressView setModalTransitionStyle:UIModalTransitionStyleCrossDissolve];
    [progressView setModalPresentationStyle:UIModalPresentationOverFullScreen];
    
    [self presentViewController:progressView animated:true completion:nil];
}

- (void)dismissProgress {
    [self dismissViewControllerAnimated:true completion:nil];
}

- (BOOL)isEmptyStr: (NSString *)string {
    if (string == nil || [string length] == 0) {
        return true;
    }
    return false;
}
@end

File: iOS/Swift/SwiftTest/SignMessage.swift
//
//  SignMessage.swift
//  SwiftTest
//
//  Created by Shilei Mao on 04/08/2021.
//

import UIKit
import HorizonFintexSDK


class SignMessage: UIViewController {
    
    @IBOutlet weak var tfKeystore: UITextField!
    
    @IBOutlet weak var tfPwd: UITextField!
    
    @IBOutlet weak var tfContractAddr: UITextField!
    @IBOutlet weak var tfMessage: UITextField!
    
    
    var ethOperator: HGEthereumOperator!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let jsEvaluator = JSWebview(handlers: nil)
        self.ethOperator = HGEthereumOperator(jsEvaluator: jsEvaluator, attachedView: self.view)
        
    }
    
    @IBAction func btSignMessageTapped(_ sender: Any) {
        guard let keystoreStr = self.tfKeystore.text?.trimmingCharacters(in: .whitespacesAndNewlines), !keystoreStr.isEmpty else {
            self.showMessage(title: "Error", message: "Please enter keystore")
            return
        }
        
        guard let pwdStr = self.tfPwd.text, !pwdStr.isEmpty else {
            self.showMessage(title: "Error", message: "Please enter password")
            return
        }
        
        guard let messageToSign = self.tfMessage.text, !messageToSign.isEmpty else {
            self.showMessage(title: "Error", message: "Please enter message")
            return
        }
        
        guard let contractAddress = self.tfContractAddr.text?.trimmingCharacters(in: .whitespacesAndNewlines), !contractAddress.isEmpty else {
            self.showMessage(title: "Error", message: "Please enter contract addr")
            return
        }
        
        guard let keystore = EthereumKeystore(jsonStr: keystoreStr) else {
            self.showMessage(title: "Error", message: "You've entered invalid keystore, please try again")
            return
        }
        
        self.showProgress()
        
        self.ethOperator.signMessage(contractAddr: contractAddress, keystore: keystore, password: pwdStr, signOriginal: messageToSign) { result, error in
            self.dismissProgress()
            
            if let error = error {
                self.showMessage(title: "Error", message: error.localizedDescription)
                return
            }
            
            guard let result = result else {
                self.showMessage(title: "Error", message: "failed to sign message, reason unknown")
                return
            }
            
            self.showMessage(title: "Signed message", message: result)
        }
        
    }
    
    func showMessage(title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let cancelAction = UIAlertAction(title: "Ok", style: .cancel, handler: nil)
        
        alert.addAction(cancelAction)
        self.present(alert, animated: true, completion: nil)
    }
    
    func showProgress() {
        if let progressView = storyboard?.instantiateViewController(identifier: "ProgressView") {
            progressView.modalTransitionStyle = .crossDissolve
            progressView.modalPresentationStyle = .overCurrentContext
            
            self.present(progressView, animated: true, completion: nil)
        }
    }
    
    func dismissProgress() {
        self.dismiss(animated: true, completion: nil)
    }
}
